internal/integration/api/watchdog.go

```diff
@@ -0,0 +1,123 @@
+
+	"github.com/siderolabs/talos/internal/integration/base"
+	"github.com/siderolabs/talos/pkg/machinery/client"
+	"github.com/siderolabs/talos/pkg/machinery/config/types/runtime"
+	runtimeres "github.com/siderolabs/talos/pkg/machinery/resources/runtime"
+)
+
+// WatchdogSuite ...
+type WatchdogSuite struct {
+	base.APISuite
+
+	ctx       context.Context //nolint:containedctx
+	ctxCancel context.CancelFunc
+}
+
+// SuiteName ...
+func (suite *WatchdogSuite) SuiteName() string {
+	return "api.WatchdogSuite"
+}
+
+// SetupTest ...
+func (suite *WatchdogSuite) SetupTest() {
+	suite.ctx, suite.ctxCancel = context.WithTimeout(context.Background(), 1*time.Minute)
+
+	if suite.Cluster == nil || suite.Cluster.Provisioner() != "qemu" {
+		suite.T().Skip("skipping watchdog test since provisioner is not qemu")
+	}
+}
+
+// TearDownTest ...
+func (suite *WatchdogSuite) TearDownTest() {
+	if suite.ctxCancel != nil {
+		suite.ctxCancel()
+	}
+}
+
+func (suite *WatchdogSuite) readWatchdogSysfs(nodeCtx context.Context, watchdog, property string) string { //nolint:unparam
+	r, err := suite.Client.Read(nodeCtx, filepath.Join("/sys/class/watchdog", watchdog, property))
+	suite.Require().NoError(err)
+
+	value, err := io.ReadAll(r)
+	suite.Require().NoError(err)
+
+	suite.Require().NoError(r.Close())
+
+	return string(bytes.TrimSpace(value))
+}
+
+// TestWatchdogSysfs sets up the watchdog and validates its parameters from the /sys/class/watchdog.
+func (suite *WatchdogSuite) TestWatchdogSysfs() {
+	// pick up a random node to test the watchdog on, and use it throughout the test
+	node := suite.RandomDiscoveredNodeInternalIP()
+
+	suite.T().Logf("testing watchdog on node %s", node)
+
+	// build a Talos API context which is tied to the node
+	nodeCtx := client.WithNode(suite.ctx, node)
+
+	// pick a watchdog
+	const watchdog = "watchdog0"
+
+	cfgDocument := runtime.NewWatchdogTimerV1Alpha1()
+	cfgDocument.WatchdogDevice = "/dev/watchdog0"
+	cfgDocument.WatchdogTimeout = 120 * time.Second
+
+	// deactivate the watchdog
+	suite.RemoveMachineConfigDocuments(nodeCtx, cfgDocument.MetaKind)
+
+	wdState := suite.readWatchdogSysfs(nodeCtx, watchdog, "state")
+	suite.Require().Equal("inactive", wdState)
+
+	suite.PatchMachineConfig(nodeCtx, cfgDocument)
+
+	_, err := suite.Client.COSI.WatchFor(nodeCtx, runtimeres.NewWatchdogTimerStatus(runtimeres.WatchdogTimerConfigID).Metadata(), state.WithEventTypes(state.Created, state.Updated))
+	suite.Require().NoError(err)
+
+	wdState = suite.readWatchdogSysfs(nodeCtx, watchdog, "state")
+	suite.Require().Equal("active", wdState)
+
+	wdTimeout := suite.readWatchdogSysfs(nodeCtx, watchdog, "timeout")
+	suite.Require().Equal("120", wdTimeout)
+
+	cfgDocument.WatchdogTimeout = 60 * time.Second
+	suite.PatchMachineConfig(nodeCtx, cfgDocument)
+
+	_, err = suite.Client.COSI.WatchFor(nodeCtx, runtimeres.NewWatchdogTimerStatus(runtimeres.WatchdogTimerConfigID).Metadata(), state.WithEventTypes(state.Created, state.Updated))
```

Comment:
It won't work this way, as the resource is already present, so you want to watch for a resource to have expected state, that is timeout of 60 seconds

Reply:
isn't state.Updated expected to catch this?

Reply:
first of all, you have both there; second it's always a race - did you start watching before it got updated or after, you never know; does any other controller update the resource (e.g. put a finalizer on it)

what you need here is: https://pkg.go.dev/github.com/cosi-project/runtime@v0.4.1/pkg/state#WithCondition

```go
state.WithCondition(func (r resource.Resource) bool {
   return r.(*runtime.WatchdogTimeStatus).TypedSpec().Timeout == 2 * time.Minute
})
```

Reply:
(wrong timeout, but I hope my point is clear)